### PR TITLE
cmake: Fix fallback quoting

### PR DESCRIPTION
**Pull request by carlocaione**
When no 'zephyr,memory-region' property is present, the region name
generation falls back to the node path. The problem is that the node
path contains the '@' character that must be quoted in the linker script
to not hit the error:

  linker_zephyr_pre0.cmd:7: ignoring invalid character `@' in expression
  linker_zephyr_pre0.cmd:7: syntax error

Add the missing quotes.

Signed-off-by: Carlo Caione <ccaione@baylibre.com>

### CODE REVIEW

File: cmake/extensions.cmake [hunk 1]
```diff
@@ -3379,7 +3379,7 @@ function(zephyr_linker_dts_memory)
   dt_prop(name PATH ${DTS_MEMORY_PATH} PROPERTY "memory-region")
   if (NOT DEFINED name)
     # Fallback to the node path
-    set(name ${DTS_MEMORY_PATH})
+    set(name \\\"${DTS_MEMORY_PATH}\\\")
```
Comment (tejlmand) [inline cmake/extensions.cmake:3382]: Maybe quoting is better done at the linker script generator itself:
https://github.com/zephyrproject-rtos/zephyr/blob/1f87642f083a7cee0aedad4bc349684ef87b1e71/cmake/linker/ld/ld_script.cmake#L36

Arm Compiler (and future toolchains with custom linkers) could need another way to handle this.

Another alternative could be to sanitize the region names, here using `zephyr_string(SANITIZE name ${name})`
https://github.com/zephyrproject-rtos/zephyr/blob/58234d490c32811860ffc818d662730448c713a7/cmake/extensions.cmake#L2285-L2302

and then similar here:
https://github.com/zephyrproject-rtos/zephyr/blob/1f87642f083a7cee0aedad4bc349684ef87b1e71/include/linker/devicetree_regions.h#L36
sanitize the node path.
@JordanYates any thoughts on this ?

**Review by tejlmand — CHANGES_REQUESTED** [review]
Good catch.
Didn't see that one coming.

**Comment by carlescufi** [comment]
@carlocaione ping

**Comment by carlocaione** [comment]
> @carlocaione ping

Waiting for @JordanYates feedback on this (https://github.com/zephyrproject-rtos/zephyr/pull/41055#discussion_r766610739)

File: cmake/extensions.cmake (hunk 1 above)
Comment (JordanYates) [reply cmake/extensions.cmake:3382]: This seems like a sane solution, and shouldn't surprise anyone given the same treatment is done with other devicetree properties.

Comment (carlocaione) [reply cmake/extensions.cmake]: It's almost xmas so I took the quickest fix :)

**Comment by carlocaione** [comment]
> @carlocaione and he responded here [#41055 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/41055#discussion_r774208969)

Yeah, I fixed it the easy way at the time. Is that ok?

**Comment by tejlmand** [comment]
> Yeah, I fixed it the easy way at the time. Is that ok?

it was, if there was a rush for having a solution merged before Christmas, but are we in rush now ?
If not, I would like to followup on this topic.

Personally I think it is safer to sanitize the name, cause then we can ensure it's identical across toolchains / linker scripts.

That also follows how paths are sanitized in the devicetree header for looking up info, for example
```
/* Node's full path: */                                                                                          
#define DT_N_S_soc_S_qspi_40029000_PATH "/soc/qspi@40029000"   
```

but if you have reasons why the qouting approach is better, please share.
I'm not fixed on sanitizing, if there are good arguments / reasons not to do it that way.

**Comment by carlocaione** [comment]
No real reasons, I was just being lazy. I'll rework this.

**Comment by carlocaione** [comment]
@tejlmand on a second thought I'm not sure if this is going to be a problem anymore when #42008 will be merged.

In that PR the `zephyr,memory-region` property is required when generating the memory region in the linker script https://github.com/zephyrproject-rtos/zephyr/blob/f32865921c272157083dd16858f1e5db40cadd94/dts/bindings/base/zephyr%2Cmemory-region.yaml#L10-L18

So this problem should be no more.

**Comment by tejlmand** [comment]
> @tejlmand on a second thought I'm not sure if this is going to be a problem anymore when #42008 will be merged.

only if we have restrictions on the name.
Do we have restrictions ?
If not, then maybe we should save ourselves from troubles in future by only allowing `[a-zA-Z0-9]` in names.

**Comment by carlocaione** [comment]
Closing this one in favor of #43877

**Pull request closed without merging**